### ISSUE TITLE
Using rocm to compile kokkos has an undefined error related to meam

### ISSUE THREAD

**Issue by DirtRain** [lammps/lammps#3862]
**Summary**

I use rocm to compile on a machine with an AMD graphics card, and I encountered an undefined error when generating the executable file of lammps.
`[100%] Linking CXX executable lmp_mpi`
`liblammps_mpi.so.0: undefined reference to 'LAMMPS_NS::PairMEAMMSKokkos<Kokkos::Serial>::PairMEAMMSKokkos(LAMMPS_NS::LAMMPS*)'`
`liblammps_mpi.so.0: undefined reference to 'LAMMPS_NS::PairMEAMKokkos<Kokkos::Serial>::PairMEAMKokkos(LAMMPS_NS::LAMMPS*)'`
`clang-14: error: linker command failed with exit code 1 (use -v to see invocation)`

**LAMMPS Version and Platform**
Lammps version is lammps-15Jun2023
Linux mymachine 3.10.0-957.nfs.5.x86_64
Gnu compiler is devtoolset_7.3.1
The rocm version is 4.0.1, but I think other versions should be able to reproduce this problem

**Further Information, Files, and Links**
The cmake command is
`
cmake \
-C ../cmake/presets/hip.cmake \
-C ../cmake/presets/most.cmake \
-C ../cmake/presets/hip_amd.cmake \
-DLAMMPS_MACHINE=mpi \
-DBUILD_SHARED_LIBS=on \
-DBUILD_MPI=yes \
-DBUILD_TOOLS=yes \
-DFFT=FFTW3 \
-DFFTW3_LIBRARY=/public/software/mathlib/fftw/gnu/lib/libfftw3.so \
-DFFTW3_INCLUDE_DIR=/public/software/fftw/gnu/include \
-DCMAKE_CXX_COMPILER=hipcc \
-DKokkos_ENABLE_HIP=yes \
-DKokkos_ARCH_VEGA906=yes \
-DPKG_KOKKOS=yes \
-DKokkos_ARCH_VEGA906=yes \
-DHIP_ARCH=gfx906 \
-DPKG_KSPACE=ON \
-DLAMMPS_INSTALL_RPATH=ON \
-DKokkos_CXX_STANDARD=17 \
-DKokkos_ENABLE_HWLOC=ON \
-DKokkos_HWLOC_DIR=/opt/hpc/software/mpi/hwloc \
../cmake/ \
-DMPI_CXX_COMPILER=mpic++ \
-DBUILD_OMP=yes \
-DCMAKE_INSTALL_PREFIX=/work/software/lammps_kokkos/install
`
I found that the rocm-related files may lack some definitions, which caused problems in my execution program generation.Is there any recent plan to fix this problem?

Thanks.

**Comment by stanmoore1** [comment]
Hmmm, this compiles successfully for me on OLCF Frontier. I cannot reproduce the link errors.

cmake -C ../cmake/presets/hip.cmake  -C ../cmake/presets/most.cmake  -C ../cmake/presets/hip_amd.cmake  -DLAMMPS_MACHINE=mpi \ -DBUILD_SHARED_LIBS=on -DBUILD_MPI=yes -DBUILD_TOOLS=yes -DCMAKE_CXX_COMPILER=hipcc -DKokkos_ENABLE_HIP=yes -DKokkos_ARCH_VEGA90A=yes -DPKG_KOKKOS=yes -DKokkos_ARCH_VEGA90A=yes -DHIP_ARCH=gfx90a -DPKG_KSPACE=ON -DLAMMPS_INSTALL_RPATH=ON -DKokkos_CXX_STANDARD=17 ../cmake/ -DMPI_CXX_COMPILER=mpic++ -DBUILD_OMP=yes

**Comment by stanmoore1** [comment]
This simplified CMake command works for me (no link errors):

cmake -C ../cmake/presets/kokkos-hip.cmake -DLAMMPS_MACHINE=mpi -DBUILD_SHARED_LIBS=on -DBUILD_MPI=yes -DCMAKE_CXX_COMPILER=hipcc -DKokkos_ENABLE_HIP=yes -DKokkos_ARCH_VEGA90A=yes -DPKG_KOKKOS=yes -DPKG_MANYBODY=yes -DKokkos_ARCH_VEGA90A=yes -DHIP_ARCH=gfx90a -DPKG_KSPACE=ON -DLAMMPS_INSTALL_RPATH=ON -DKokkos_CXX_STANDARD=17 ../cmake/ -DMPI_CXX_COMPILER=mpic++

```
./lmp_mpi -h
...
Compiler: Clang C++ AMD Clang 15.0.0 (https://github.com/RadeonOpenCompute/llvm-
project roc-5.4.3 23045 a29fe425c7b0e5aba97ed2f95f61fd5ecba68aed) with OpenMP 5.
0
C++ standard: C++17
MPI v3.1: MPI VERSION    : CRAY MPICH version 8.1.23.14 (ANL base 3.4a2)
MPI BUILD INFO : Tue Nov 29 12:34 2022 (git hash 210ae8b)

Accelerator configuration:

KOKKOS package API: HIP OpenMP Serial
KOKKOS package precision: double
...
Installed packages:

KOKKOS KSPACE MANYBODY 
```

**Comment by stanmoore1** [comment]
This more complex CMake command also works for me on Frontier:

cmake -C ../cmake/presets/kokkos-hip.cmake -C ../cmake/presets/most.cmake -DLAMMPS_MACHINE=mpi -DBUILD_SHARED_LIBS=on -DBUILD_TOOLS=no -DBUILD_MPI=yes -DCMAKE_CXX_COMPILER=hipcc -DKokkos_ENABLE_HIP=yes -DKokkos_ARCH_VEGA90A=yes -DPKG_KOKKOS=yes -DPKG_PHONON=no -DKokkos_ARCH_VEGA90A=yes -DHIP_ARCH=gfx90a -DPKG_KSPACE=ON -DLAMMPS_INSTALL_RPATH=ON -DKokkos_CXX_STANDARD=17 ../cmake/ -DMPI_CXX_COMPILER=mpic++

**Comment by DirtRain** [comment]
@stanmoore1 Thank you for your reply.
There is a problem with the rocm information I gave at the beginning. In fact, the rocm version I use is 5.1. I am sorry that I gave the wrong information.
Then I tested the command you gave, but it didn't work very well on my machine.I want to confirm that the version of the software you compiled is released in 2023?
I found that the KOKKOS support for the `MEAM` potential function only started in 2023.  Compilation before 2023 is no problem.The version released in 2023 has more files such as `meam_kokkos.h` in the `src/KOKKOS` folder. I guess these files are what's causing me to fail, but these files are also what I need.

Thanks again for taking the time to answer my question

**Comment by stanmoore1** [comment]
I'm using latest develop. Since it works on my system but not yours, we need to find out what the difference is. Can you please simplify your CMake command to the smallest possible to reproduce the error? In other words do not use `../cmake/presets/most.cmake` and get rid of everything possible. Thanks

This is the version of LAMMPS I'm using:
```
commit 80f3b6699ef4b2303a2d03856864df514d3fd3ed (HEAD -> develop)
Merge: e27895b232 03b25dcbaa
Author: Axel Kohlmeyer <akohlmey@gmail.com>
Date:   Wed Jun 14 12:29:07 2023 -0400

    Merge pull request #3817 from jrgissing/bond/react-charge_rescale+other_small_fixes
```

**Comment by stanmoore1** [comment]
Actually I think I see the issue here:
https://github.com/lammps/lammps/blob/develop/src/KOKKOS/pair_meam_kokkos.cpp#L983

Not sure why it works for me, I will submit a PR.

**Comment by stanmoore1** [comment]
@DirtRain this should be fixed by #3864, can you confirm?
